File: src/main/java/pl/mjedynak/idea/plugins/builder/factory/PsiFieldsForBuilderFactory.java
package pl.mjedynak.idea.plugins.builder.factory;

import com.google.common.collect.Lists;
import com.intellij.codeInsight.generation.PsiElementClassMember;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiField;
import com.intellij.psi.PsiMethod;
import java.util.List;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.builder.psi.BestConstructorSelector;
import pl.mjedynak.idea.plugins.builder.psi.model.PsiFieldsForBuilder;
import pl.mjedynak.idea.plugins.builder.verifier.PsiFieldVerifier;

public class PsiFieldsForBuilderFactory {

    private PsiFieldsForBuilderFactory() {
        throw new UnsupportedOperationException("Utility class");
    }

    @Contract("_, _ -> new")
    public static @NotNull PsiFieldsForBuilder createPsiFieldsForBuilder(
            @NotNull List<PsiElementClassMember<?>> psiElementClassMembers, @NotNull PsiClass psiClass) {
        List<PsiField> allSelectedPsiFields = Lists.newArrayList();
        List<PsiField> psiFieldsFoundInSetters = Lists.newArrayList();
        for (PsiElementClassMember<?> psiElementClassMember : psiElementClassMembers) {
            PsiElement psiElement = psiElementClassMember.getPsiElement();
            if (psiElement instanceof PsiField) {
                allSelectedPsiFields.add((PsiField) psiElement);
                if (PsiFieldVerifier.isSetInSetterMethod((PsiField) psiElement, psiClass)) {
                    psiFieldsFoundInSetters.add((PsiField) psiElement);
                }
            }
        }
        List<PsiField> psiFieldsToFindInConstructor = getSubList(allSelectedPsiFields, psiFieldsFoundInSetters);
        List<PsiField> psiFieldsForConstructor = Lists.newArrayList();
        PsiMethod bestConstructor = BestConstructorSelector.getBestConstructor(psiFieldsToFindInConstructor, psiClass);
        if (bestConstructor != null) {
            buildPsiFieldsForConstructor(psiFieldsForConstructor, allSelectedPsiFields, bestConstructor);
        }
        List<PsiField> psiFieldsForSetters = getSubList(psiFieldsFoundInSetters, psiFieldsForConstructor);

        return new PsiFieldsForBuilder(
                psiFieldsForSetters, psiFieldsForConstructor, allSelectedPsiFields, bestConstructor);
    }

    private static void buildPsiFieldsForConstructor(
            @NotNull List<PsiField> psiFieldsForConstructor,
            @NotNull List<PsiField> allSelectedPsiFields,
            @NotNull PsiMethod bestConstructor) {
        for (PsiField selectedPsiField : allSelectedPsiFields) {
            if (PsiFieldVerifier.checkConstructor(selectedPsiField, bestConstructor)) {
                psiFieldsForConstructor.add(selectedPsiField);
            }
        }
    }

    private static @NotNull List<PsiField> getSubList(
            @NotNull List<PsiField> inputList, @NotNull List<PsiField> listToRemove) {
        List<PsiField> newList = Lists.newArrayList();
        for (PsiField inputPsiField : inputList) {
            boolean setterMustBeAdded = true;
            for (PsiField psiFieldToRemove : listToRemove) {
                if (psiFieldToRemove.getName().equals(inputPsiField.getName())) {
                    setterMustBeAdded = false;
                }
            }
            if (setterMustBeAdded) {
                newList.add(inputPsiField);
            }
        }
        return newList;
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/factory/PsiFieldsForBuilderFactory.kt
package io.github.kingg22.kobra.builder.factory

import com.intellij.codeInsight.generation.PsiElementClassMember
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiField
import com.intellij.psi.PsiMethod
import io.github.kingg22.kobra.builder.psi.BestConstructorSelector
import io.github.kingg22.kobra.builder.psi.model.PsiFieldsForBuilder
import io.github.kingg22.kobra.builder.verifier.PsiFieldVerifier
import org.jetbrains.annotations.Contract

public object PsiFieldsForBuilderFactory {
    @Contract("_, _ -> new")
    @JvmStatic
    public fun createPsiFieldsForBuilder(
        psiElementClassMembers: List<PsiElementClassMember<*>>,
        psiClass: PsiClass,
    ): PsiFieldsForBuilder {
        val allSelectedPsiFields: MutableList<PsiField> = mutableListOf()
        val psiFieldsFoundInSetters: MutableList<PsiField> = mutableListOf()
        for (psiElementClassMember in psiElementClassMembers) {
            val psiElement = psiElementClassMember.getPsiElement()
            if (psiElement is PsiField) {
                allSelectedPsiFields.add(psiElement)
                if (PsiFieldVerifier.isSetInSetterMethod(psiElement, psiClass)) {
                    psiFieldsFoundInSetters.add(psiElement)
                }
            }
        }
        val psiFieldsToFindInConstructor = getSubList(allSelectedPsiFields, psiFieldsFoundInSetters)
        val psiFieldsForConstructor: MutableList<PsiField> = mutableListOf()
        val bestConstructor = BestConstructorSelector.getBestConstructor(psiFieldsToFindInConstructor, psiClass)
        if (bestConstructor != null) {
            buildPsiFieldsForConstructor(psiFieldsForConstructor, allSelectedPsiFields, bestConstructor)
        }
        val psiFieldsForSetters = getSubList(psiFieldsFoundInSetters, psiFieldsForConstructor)

        return PsiFieldsForBuilder(
            psiFieldsForSetters,
            psiFieldsForConstructor,
            allSelectedPsiFields,
            bestConstructor,
        )
    }

    private fun buildPsiFieldsForConstructor(
        psiFieldsForConstructor: MutableList<PsiField>,
        allSelectedPsiFields: List<PsiField>,
        bestConstructor: PsiMethod,
    ) {
        for (selectedPsiField in allSelectedPsiFields) {
            if (PsiFieldVerifier.checkConstructor(selectedPsiField, bestConstructor)) {
                psiFieldsForConstructor.add(selectedPsiField)
            }
        }
    }

    private fun getSubList(inputList: List<PsiField>, listToRemove: List<PsiField>): List<PsiField> {
        val newList: MutableList<PsiField> = mutableListOf()
        for (inputPsiField in inputList) {
            var setterMustBeAdded = true
            for (psiFieldToRemove in listToRemove) {
                if (psiFieldToRemove.name == inputPsiField.name) {
                    setterMustBeAdded = false
                }
            }
            if (setterMustBeAdded) {
                newList.add(inputPsiField)
            }
        }
        return newList
    }
}
